Added TimePicker TimeSelected Event

Axemasta opened the pull request:
### Description of Change

The TimePicker control does not have an event raised specifically when the Time property changes. Whilst you can subscribe to the general `PropertyChanged` event and filter out the time property:
```csharp
private void OnTimePickerPropertyChanged(object? sender, PropertyChangedEventArgs e)
{
	if (string.IsNullOrEmpty(e.PropertyName))
	{
		return;
	}

	if (e.PropertyName.Equals(TimePicker.TimeProperty.PropertyName, StringComparison.Ordinal))
	{
		// Do Something
	}
}
```
This is not very elegant and not inline with many other controls in the maui suit, mainly with its partner `DatePicker` which has the `DateSelected` event.

This PR adds `TimeSelected` to the `TimePicker`, following the paradigm seen with `DatePicker`.

I have updated the documentation for `TimeChangedEventArgs` inline with `DateChangedEventArgs`, I'm not sure how these get generated so I did it by hand. Any help / pointers to format this I am all ears!

### Issues Fixed

None reported, but the event is missing. I figured just raising the PR with the change was easier!

Fixes #

`TimePicker` is now easier to build controls & logic around 🥳

Axemasta committed:
Added TimeSelected Event To TimePicker

Currently there is no event to subscribe to for TimePicker changed events, You have to use the PropertyChanged event and filter by property name. This is a nicer way of subscribing to said event.

Axemasta committed:
Added TimeSelected Unit Tests

Copying the same pattern from DatePicker & updating for the TimePicker

mattleibow commented on src/Controls/src/Core/TimePicker/TimePicker.cs:
```diff
@@ -157,5 +160,14 @@ TimeSpan ITimePicker.Time
+
+		static void TimePropertyChanged(BindableObject bindable, object oldValue, object newValue)
+		{
+			var timePicker = (TimePicker)bindable;
+			EventHandler<TimeChangedEventArgs> selected = timePicker.TimeSelected;
+
+			if (selected != null)
+				selected(timePicker, new TimeChangedEventArgs((TimeSpan)oldValue, (TimeSpan)newValue));
```
```suggestion
			if (bindable is TimePicker timePicker)
				timePicker.TimeSelected?.Invoke(timePicker, new TimeChangedEventArgs((TimeSpan)oldValue, (TimeSpan)newValue));
```

mattleibow commented on src/Controls/tests/Core.UnitTests/TimePickerUnitTest.cs:
```diff
@@ -39,5 +39,57 @@ public void ZeroTimeIsValid()
+		{
+			var picker = new TimePicker();
+
+			bool selected = false;
```
Instead of a `bool`, could this be an `int` that is incremented in the event? This way the Assert can also verify that the event was fired only once.

Axemasta replied:
Updated, great idea!

mattleibow commented on src/Controls/src/Core/TimeChangedEventArgs.cs:
```diff
@@ -0,0 +1,22 @@
+
+namespace Microsoft.Maui.Controls
+{
+	/// <include file="../../docs/Microsoft.Maui.Controls/TimeChangedEventArgs.xml" path="Type[@FullName='Microsoft.Maui.Controls.TimeChangedEventArgs']/Docs/*" />
```
Thanks for the docs! The new way going forward is to rather put the docs directly in the code. Could you move the docs out of the separate xml file (and delete) and into the cs files?

Axemasta replied:
I've moved them, I wasn't sure what to do with the Remarks & Value tags so I have preserved them. I also don't know if you prefer all on 1 line:
```xml
<summary>The time that the user entered.</summary>
```
or tags seperated by line:
```xml
<summary>
The time that the user entered.
</summary>
```

mattleibow replied:
Multi-line is fine, but we do a mix so either way does not matter.

mattleibow reviewed (commented):
Thanks for this nice PR, code, docs and tests! Just have a few comments on some small changes, but otherwise looking good!

Axemasta committed:
Code Review Suggestions

Moved docs for TimeChangedEventArgs to file as per request. Refactored bindable property code to be alot nicer. Updated unit tests to test for multiple executions

mattleibow commented:
/azp run

jfversluis reviewed (dismissed):
New APIs need to be added to the right txt files. Looks like you added it for netstandard, but not the other targets

Let us know if you can't figure it out @Axemasta 

> D:\a\_work\1\s\src\Controls\src\Core\TimePicker\TimePicker.cs(111,51): error RS0016: Symbol 'TimeSelected' is not part of the declared public API (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) [D:\a\_work\1\s\src\Controls\src\Core\Controls.Core.csproj::TargetFramework=net8.0-android]
D:\a\_work\1\s\src\Controls\src\Core\TimeChangedEventArgs.cs(8,15): error RS0016: Symbol 'TimeChangedEventArgs' is not part of the declared public API (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) [D:\a\_work\1\s\src\Controls\src\Core\Controls.Core.csproj::TargetFramework=net8.0-android]
D:\a\_work\1\s\src\Controls\src\Core\TimeChangedEventArgs.cs(14,10): error RS0016: Symbol 'TimeChangedEventArgs' is not part of the declared public API (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) [D:\a\_work\1\s\src\Controls\src\Core\Controls.Core.csproj::TargetFramework=net8.0-android]
D:\a\_work\1\s\src\Controls\src\Core\TimeChangedEventArgs.cs(23,29): error RS0016: Symbol 'NewTime.get' is not part of the declared public API (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) [D:\a\_work\1\s\src\Controls\src\Core\Controls.Core.csproj::TargetFramework=net8.0-android]
D:\a\_work\1\s\src\Controls\src\Core\TimeChangedEventArgs.cs(28,29): error RS0016: Symbol 'OldTime.get' is not part of the declared public API (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) [D:\a\_work\1\s\src\Controls\src\Core\Controls.Core.csproj::TargetFramework=net8.0-android]
D:\a\_work\1\s\src\Controls\src\Core\TimeChangedEventArgs.cs(8,15): error RS0016: Symbol 'TimeChangedEventArgs' is not part of the declared public API (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) [D:\a\_work\1\s\src\Controls\src\Core\Controls.Core.csproj::TargetFramework=net8.0-windows10.0.20348]
D:\a\_work\1\s\src\Controls\src\Core\TimePicker\TimePicker.cs(111,51): error RS0016: Symbol 'TimeSelected' is not part of the declared public API (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) [D:\a\_work\1\s\src\Controls\src\Core\Controls.Core.csproj::TargetFramework=net8.0-windows10.0.20348]
D:\a\_work\1\s\src\Controls\src\Core\TimeChangedEventArgs.cs(14,10): error RS0016: Symbol 'TimeChangedEventArgs' is not part of the declared public API (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) [D:\a\_work\1\s\src\Controls\src\Core\Controls.Core.csproj::TargetFramework=net8.0-windows10.0.20348]
D:\a\_work\1\s\src\Controls\src\Core\TimeChangedEventArgs.cs(23,29): error RS0016: Symbol 'NewTime.get' is not part of the declared public API (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) [D:\a\_work\1\s\src\Controls\src\Core\Controls.Core.csproj::TargetFramework=net8.0-windows10.0.20348]
D:\a\_work\1\s\src\Controls\src\Core\TimeChangedEventArgs.cs(28,29): error RS0016: Symbol 'OldTime.get' is not part of the declared public API (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) [D:\a\_work\1\s\src\Controls\src\Core\Controls.Core.csproj::TargetFramework=net8.0-windows10.0.20348]
D:\a\_work\1\s\src\Controls\src\Core\TimePicker\TimePicker.cs(111,51): error RS0016: Symbol 'TimeSelected' is not part of the declared public API (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) [D:\a\_work\1\s\src\Controls\src\Core\Controls.Core.csproj::TargetFramework=net8.0-tizen]
D:\a\_work\1\s\src\Controls\src\Core\TimeChangedEventArgs.cs(8,15): error RS0016: Symbol 'TimeChangedEventArgs' is not part of the declared public API (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) [D:\a\_work\1\s\src\Controls\src\Core\Controls.Core.csproj::TargetFramework=net8.0-tizen]
D:\a\_work\1\s\src\Controls\src\Core\TimeChangedEventArgs.cs(14,10): error RS0016: Symbol 'TimeChangedEventArgs' is not part of the declared public API (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) [D:\a\_work\1\s\src\Controls\src\Core\Controls.Core.csproj::TargetFramework=net8.0-tizen]
D:\a\_work\1\s\src\Controls\src\Core\TimeChangedEventArgs.cs(23,29): error RS0016: Symbol 'NewTime.get' is not part of the declared public API (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) [D:\a\_work\1\s\src\Controls\src\Core\Controls.Core.csproj::TargetFramework=net8.0-tizen]
D:\a\_work\1\s\src\Controls\src\Core\TimeChangedEventArgs.cs(28,29): error RS0016: Symbol 'OldTime.get' is not part of the declared public API (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) [D:\a\_work\1\s\src\Controls\src\Core\Controls.Core.csproj::TargetFramework=net8.0-tizen]
D:\a\_work\1\s\src\Controls\src\Core\TimePicker\TimePicker.cs(111,51): error RS0016: Symbol 'TimeSelected' is not part of the declared public API (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) [D:\a\_work\1\s\src\Controls\src\Core\Controls.Core.csproj::TargetFramework=net8.0]
D:\a\_work\1\s\src\Controls\src\Core\TimeChangedEventArgs.cs(8,15): error RS0016: Symbol 'TimeChangedEventArgs' is not part of the declared public API (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) [D:\a\_work\1\s\src\Controls\src\Core\Controls.Core.csproj::TargetFramework=net8.0]
D:\a\_work\1\s\src\Controls\src\Core\TimeChangedEventArgs.cs(14,10): error RS0016: Symbol 'TimeChangedEventArgs' is not part of the declared public API (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) [D:\a\_work\1\s\src\Controls\src\Core\Controls.Core.csproj::TargetFramework=net8.0]
D:\a\_work\1\s\src\Controls\src\Core\TimeChangedEventArgs.cs(23,29): error RS0016: Symbol 'NewTime.get' is not part of the declared public API (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) [D:\a\_work\1\s\src\Controls\src\Core\Controls.Core.csproj::TargetFramework=net8.0]
D:\a\_work\1\s\src\Controls\src\Core\TimeChangedEventArgs.cs(28,29): error RS0016: Symbol 'OldTime.get' is not part of the declared public API (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) [D:\a\_work\1\s\src\Controls\src\Core\Controls.Core.csproj::TargetFramework=net8.0]
D:\a\_work\1\s\src\Controls\src\Core\TimePicker\TimePicker.cs(111,51): error RS0016: Symbol 'TimeSelected' is not part of the declared public API (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) [D:\a\_work\1\s\src\Controls\src\Core\Controls.Core.csproj::TargetFramework=net8.0-windows10.0.19041]
D:\a\_work\1\s\src\Controls\src\Core\TimeChangedEventArgs.cs(8,15): error RS0016: Symbol 'TimeChangedEventArgs' is not part of the declared public API (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) [D:\a\_work\1\s\src\Controls\src\Core\Controls.Core.csproj::TargetFramework=net8.0-windows10.0.19041]
D:\a\_work\1\s\src\Controls\src\Core\TimeChangedEventArgs.cs(14,10): error RS0016: Symbol 'TimeChangedEventArgs' is not part of the declared public API (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) [D:\a\_work\1\s\src\Controls\src\Core\Controls.Core.csproj::TargetFramework=net8.0-windows10.0.19041]
D:\a\_work\1\s\src\Controls\src\Core\TimeChangedEventArgs.cs(23,29): error RS0016: Symbol 'NewTime.get' is not part of the declared public API (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) [D:\a\_work\1\s\src\Controls\src\Core\Controls.Core.csproj::TargetFramework=net8.0-windows10.0.19041]
D:\a\_work\1\s\src\Controls\src\Core\TimeChangedEventArgs.cs(28,29): error RS0016: Symbol 'OldTime.get' is not part of the declared public API (https://github.com/dotnet/roslyn-analyzers/blob/main/src/PublicApiAnalyzers/PublicApiAnalyzers.Help.md) [D:\a\_work\1\s\src\Controls\src\Core\Controls.Core.csproj::TargetFramework=net8.0-windows10.0.19041]

Axemasta commented:
@jfversluis Is there any tool I can run to assist with this process? The current additions were made with a tooltip in vs4mac & i'm not sure how the team would normally approach managing these :)

jfversluis commented:
The same as you would I'm afraid 😅 you can switch context with this combobox in VS and then run the lightbulb action for each one. (screenshot not related to your change in this case)

![image](https://github.com/dotnet/maui/assets/939291/182a10d0-a24b-4d2a-b11f-4b0ccb84fc13)

Or just do it for one, as you have now, and copy those lines over to the other txt files manually which is probably faster.

Axemasta committed:
Added TimeChangedEventArgs to public api

Couldnt do this automatically since vs didn't pick this up. I've instead done this by hand.

mattleibow commented:
/azp run

mattleibow reviewed (dismissed):
This is great! Thanks for the work and getting all the bits in.

jsuarezruiz reviewed (commented):
Should we retarget this changes to the net9.0 branch?

rmarinho commented:
Hello this needs to retarget net9.0 please since no new api can be added to existing versions.

Axemasta commented:
@rmarinho I have rebased and updated target to `net9.0`

Axemasta commented:
@samhouts This PR is retargetted and ready to reopen, however I don't have permission to reopen it. Please could you reopen this since I don't want to raise a new PR since this all the review history is here :)

mattleibow commented:
Me neither... @rmarinho can you?
![image](https://github.com/dotnet/maui/assets/1096616/522f24f4-e35d-4596-97c9-f8939269092b)

mattleibow commented:
Now I can....

mattleibow commented:
Nope, disabled again.

mattleibow commented:
One more time...

mattleibow commented:
Nope, it says I can and then disables during the process.

mattleibow commented:
If Rui can't, then might be worth just re-creating. OR recreate anyways.

Axemasta commented:
How bizarre, I don't even see the reopen button. I have reopened the PR here #21156
